py-polars/src/series/export.rs

```diff
@@ -161,111 +163,134 @@ impl PySeries {
-            Int64 => numeric_series_to_numpy::<Int64Type, f64>(py, s),
-            UInt8 => numeric_series_to_numpy::<UInt8Type, f32>(py, s),
-            UInt16 => numeric_series_to_numpy::<UInt16Type, f32>(py, s),
-            UInt32 => numeric_series_to_numpy::<UInt32Type, f64>(py, s),
-            UInt64 => numeric_series_to_numpy::<UInt64Type, f64>(py, s),
-            Float32 => numeric_series_to_numpy::<Float32Type, f32>(py, s),
-            Float64 => numeric_series_to_numpy::<Float64Type, f64>(py, s),
-            Boolean => {
-                let ca = s.bool().unwrap();
-                let np_arr = PyArray1::from_iter_bound(py, ca.into_iter().map(|s| s.into_py(py)));
-                np_arr.into_py(py)
-            },
-            Date => date_series_to_numpy(py, s),
-            Datetime(tu, _) => {
-                use numpy::datetime::{units, Datetime};
-                match tu {
-                    TimeUnit::Milliseconds => {
-                        temporal_series_to_numpy::<Datetime<units::Milliseconds>>(py, s)
-                    },
-                    TimeUnit::Microseconds => {
-                        temporal_series_to_numpy::<Datetime<units::Microseconds>>(py, s)
-                    },
-                    TimeUnit::Nanoseconds => {
-                        temporal_series_to_numpy::<Datetime<units::Nanoseconds>>(py, s)
-                    },
-                }
-            },
-            Duration(tu) => {
-                use numpy::datetime::{units, Timedelta};
-                match tu {
-                    TimeUnit::Milliseconds => {
-                        temporal_series_to_numpy::<Timedelta<units::Milliseconds>>(py, s)
-                    },
-                    TimeUnit::Microseconds => {
-                        temporal_series_to_numpy::<Timedelta<units::Microseconds>>(py, s)
-                    },
-                    TimeUnit::Nanoseconds => {
-                        temporal_series_to_numpy::<Timedelta<units::Nanoseconds>>(py, s)
-                    },
+    /// This method copies data only when necessary. Set `allow_copy` to raise an error if copy
+    /// is required. Set `writable` to make sure the resulting array is writable, possibly requiring
+    /// copying the data.
+    fn to_numpy(&self, py: Python, allow_copy: bool, writable: bool) -> PyResult<PyObject> {
+        let is_empty = self.series.is_empty();
+
+        if self.series.null_count() == 0 {
+            if let Some(mut arr) = self.to_numpy_view(py) {
+                if writable || is_empty {
+                    if !allow_copy && !is_empty {
+                        return Err(PyValueError::new_err(
+                            "cannot return a zero-copy writable array",
+                        ));
+                    }
+                    // TODO: Is there a Rust-native way to do this?
+                    arr = arr.call_method0(py, intern!(py, "copy"))?;
```

Comment:
I think this is the rust native way to copy a python object. We have to interact with python somehow.

Reply:
It's not a generic Python copy, it calls the `ndarray.copy` method:
https://numpy.org/doc/stable/reference/generated/numpy.ndarray.copy.html

I figured maybe the numpy crate might have the same functionality available somewhere, but I couldn't find it. In any case, this code should be fine - I removed the TODO comment.